### PR TITLE
Add Array#flatMap method

### PR DESCRIPTION
**Pull request by goodmind**
Fixes part of https://github.com/facebook/flow/issues/7397

### CODE REVIEW

**Comment by goodmind** [comment]
@nmote

File: tests/arraylib/array_lib.js [hunk 1]
```diff
@@ -62,3 +62,53 @@ function of_test() {
...
+  }
+  function case4() {
+    let a = [5, 2, 3, 4];
+    let b: Array<number | string> = a.flatMap(n => (n < 0 ? [1, 2, 3] : "ok"));
+    let c: $ReadOnlyArray<number> = [5, 2, 3, 4];
+    let d: Array<number | string> = a.flatMap(n => (n < 0 ? [1, 2, 3] : "ok"));
+  }
+  function case5() {
+    let c: $ReadOnlyArray<number> = [5, 2, 3, 4];
+    let d: Array<number> = a.flatMap(n => {
```
Comment (chicoxyzzy) [inline tests/arraylib/array_lib.js]: Seems that `a` is missing here

File: lib/core.js [hunk 2]
```diff
@@ -226,6 +226,7 @@ declare class $ReadOnlyArray<+T> {
     keys(): Iterator<number>;
     lastIndexOf(searchElement: mixed, fromIndex?: number): number;
     map<U>(callbackfn: (value: T, index: number, array: $ReadOnlyArray<T>) => U, thisArg?: any): Array<U>;
```
Comment (kangax) [inline lib/core.js]: why does this return `$ReadOnlyArray<U>` while map returns just `U`?

Comment (goodmind) [reply lib/core.js]: https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap

**Comment by samwgoldman** [comment]
Why does the mapper function return `$ReadOnlyArray<U> | U`, instead of just `$ReadOnlyArray<U>`?

**Comment by goodmind** [comment]
@samwgoldman flatMap can work like `map`
```js
[ [1, 2, 3] ].flatMap(() => 1) // [1]
```

**Comment by samwgoldman** [comment]
Oof. Annoying, but certainly in line with the gratuitously overloaded nature of existing APIs.

**Comment by goodmind** [comment]
Well, it's easier to use with conditionals
```
[ [1, 2, 3] ].flatMap((a) => a[0] === 2 ? a : 1)
```
this would be much bigger if `flatMap` supported only arrays

**Comment by samwgoldman** [comment]
It seems easy enough to just write this:

```
[ [1, 2, 3] ].flatMap((a) => a[0] === 2 ? a : [1])
```

**Comment by goodmind** [comment]
oops

**Pull request closed without merging**